src/evm/protocol/uniswap_v4/state.rs

```diff
@@ -1,26 +1,51 @@
-use alloy_primitives::U256;
+#![allow(dead_code)] //TODO: remove when used
 
-use crate::evm::protocol::utils::uniswap::tick_list::{TickInfo, TickList};
+use alloy_primitives::{Sign, I256, U256};
+
+use crate::{
+    evm::protocol::{
+        safe_math::{safe_add_u256, safe_sub_u256},
+        u256_num::u256_to_biguint,
+        utils::uniswap::{
+            liquidity_math, swap_math,
+            tick_list::{TickInfo, TickList, TickListErrorKind},
+            tick_math::{
+                get_sqrt_ratio_at_tick, get_tick_at_sqrt_ratio, MAX_SQRT_RATIO, MAX_TICK,
+                MIN_SQRT_RATIO, MIN_TICK,
+            },
+            StepComputation, SwapResults, SwapState,
+        },
+    },
+    protocol::{errors::SimulationError, models::GetAmountOutResult, state::ProtocolSim},
+};
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct UniswapV4State {
     liquidity: u128,
     sqrt_price: U256,
-    lp_fee: i32,
-    protocol_fees: UniswapV4ProtocolFees,
+    fees: UniswapV4Fees,
     tick: i32,
     ticks: TickList,
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub struct UniswapV4ProtocolFees {
-    zero2one: i32,
-    one2zero: i32,
+pub struct UniswapV4Fees {
+    // Protocol fees in the zero for one direction
+    zero_for_one: u32,
+    // Protocol fees in the one for zero direction
+    one_for_zero: u32,
+    // Liquidity providers fees
+    lp_fee: u32,
 }
 
-impl UniswapV4ProtocolFees {
-    pub fn new(zero2one: i32, one2zero: i32) -> Self {
-        Self { zero2one, one2zero }
+impl UniswapV4Fees {
+    pub fn new(zero_for_one: u32, one_for_zero: u32, lp_fee: u32) -> Self {
+        Self { zero_for_one, one_for_zero, lp_fee }
+    }
+
+    fn calculate_swap_fees_pips(&self, zero_for_one: bool) -> u32 {
```

Comment:
What are pips?

Reply:
It's the unit they use for fees, basically `1 pip = 0.01%`

Not sure what's the diff between pips and bips though, I've seen both in their contract

Reply:
ah ok! I've heard of bips, never pips. cool